common/network-common/src/main/java/org/apache/spark/network/sasl/SaslRpcHandler.java

```diff
@@ -80,47 +84,79 @@ public void receive(TransportClient client, ByteBuffer message, RpcResponseCallb
+        // First message in the handshake, setup the necessary state.
+        client.setClientId(saslMessage.appId);
+        saslServer = new SparkSaslServer(saslMessage.appId, secretKeyHolder,
+          conf.saslServerAlwaysEncrypt());
+      }
 
-    byte[] response;
-    try {
-      response = saslServer.response(JavaUtils.bufferToArray(
-        saslMessage.body().nioByteBuffer()));
-    } catch (IOException ioe) {
-      throw new RuntimeException(ioe);
+      byte[] response;
+      try {
+        response = saslServer.response(JavaUtils.bufferToArray(
+          saslMessage.body().nioByteBuffer()));
+      } catch (IOException ioe) {
+        throw new RuntimeException(ioe);
+      }
+      callback.onSuccess(ByteBuffer.wrap(response));
     }
-    callback.onSuccess(ByteBuffer.wrap(response));
 
     // Setup encryption after the SASL response is sent, otherwise the client can't parse the
     // response. It's ok to change the channel pipeline here since we are processing an incoming
     // message, so the pipeline is busy and no new incoming messages will be fed to it before this
     // method returns. This assumes that the code ensures, through other means, that no outbound
     // messages are being written to the channel while negotiation is still going on.
     if (saslServer.isComplete()) {
-      logger.debug("SASL authentication successful for channel {}", client);
-      isComplete = true;
-      if (SparkSaslServer.QOP_AUTH_CONF.equals(saslServer.getNegotiatedProperty(Sasl.QOP))) {
+      if (!SparkSaslServer.QOP_AUTH_CONF.equals(saslServer.getNegotiatedProperty(Sasl.QOP))) {
+        logger.debug("SASL authentication successful for channel {}", client);
+        saslServer.dispose();
+        saslServer = null;
+        isComplete = true;
+        return ;
+      }
+
+      if (!conf.saslEncryptionAesEnabled()) {
         logger.debug("Enabling encryption for channel {}", client);
         SaslEncryption.addToChannel(channel, saslServer, conf.maxSaslEncryptedBlockSize());
-        saslServer = null;
       } else {
-        saslServer.dispose();
-        saslServer = null;
+        // Extra negotiation should happen after authentication, so return directly while
+        // processing authenticate.
+        if (!isAuthenticated) {
+          logger.debug("SASL authentication successful for channel {}", client);
+          isAuthenticated = true;
+          return;
+        } else {
+          try {
+            AesConfigMessage configMessage = AesConfigMessage.decodeMessage(message);
+            configMessage = AesCipher.responseConfigMessage(configMessage);
+            AesCipher cipher = new AesCipher(configMessage);
+
+            ByteBuffer buf = configMessage.encodeMessage();
+
+            // Encrypt the config message.
+            ByteBuffer encrypted = ByteBuffer.wrap(
+              saslServer.wrap(buf.array(), 0, buf.array().length));
+            callback.onSuccess(encrypted);
```

Comment:
At this point you're not using `saslServer` anymore, so you should dispose of it. I kinda prefer my previous suggestion where instead of `if..else` you have just `if`s and `return`s. I think it makes it a little easier to follow the code.

You can abstract the `saslServer` cleanup by adding a `complete(boolean dispose)` method that updates `isComplete` and `saslServer` and is called from this method where needed.